python/taichi/lang/matrix.py

```diff
@@ -311,26 +311,91 @@ def _linearize_entry_id(self, *args):
             f"The 1-th matrix index is out of range: 0 <= {args[1]} < {self.m}"
         return args[0] * self.m + args[1]
 
+    # host access & python scope operation
+    def __len__(self):
+        """Get the length of each row of a matrix"""
+        # TODO: When this is a vector, should return its dimension?
+        return self.n
+
+    def __iter__(self):
+        if self.m == 1:
+            return (self(i) for i in range(self.n))
+        return ([self(i, j) for j in range(self.m)] for i in range(self.n))
+
+    @python_scope
+    def __getitem__(self, indices):
+        """Access to the element at the given indices in a matrix.
+
+        Args:
+            indices (Sequence[Expr]): the indices of the element.
+
+        Returns:
+            The value of the element at a specific position of a matrix.
+
+        """
+        if not isinstance(indices, (list, tuple)):
+            indices = [indices]
+        assert len(indices) in [1, 2]
+        i = indices[0]
+        j = 0 if len(indices) == 1 else indices[1]
+        if isinstance(i, slice) or isinstance(j, slice):
+            return self._get_slice(i, j)
+        return self._get_entry_and_read_if_in_pyscope([i, j])
+
+    @python_scope
+    def __setitem__(self, indices, item):
+        """Set the element value at the given indices in a matrix.
+
+        Args:
+            indices (Sequence[Expr]): the indices of a element.
+
+        """
+        if not isinstance(indices, (list, tuple)):
+            indices = [indices]
+        assert len(indices) in [1, 2]
+        i = indices[0]
+        j = 0 if len(indices) == 1 else indices[1]
+        idx = self._linearize_entry_id(i, j)
+        if isinstance(self.entries[idx], SNodeHostAccess):
+            self.entries[idx].accessor.setter(item, *self.entries[idx].key)
+        elif isinstance(self.entries[idx], NdarrayHostAccess):
+            self.entries[idx].setter(item)
+        else:
+            self.entries[idx] = item
+
     def __call__(self, *args, **kwargs):
+        # TODO: It's quite hard to search for __call__, consider replacing this
+        # with a method of actual names?
         assert kwargs == {}
-        ret = self._get_entry(*args)
+        return self._get_entry_and_read_if_in_pyscope(args)
+
+    def _get_entry_and_read_if_in_pyscope(self, indices):
+        # Can be invoked in both Python and Taichi scope. `indices` must be
+        # compile-time constants (e.g. Python values)
+        ret = self._get_entry(*indices)
+
+        def ensure_pyscope():
```

Comment:
Is this necessary? IIUC, `SNodeHostAccess` and `NdarrayHostAccess` are created only when accessing a matrix field/ndarray in python scope, and they never appear in Taichi scope.

Reply:
Alright, it's possible to be passed in as a template argument, and it goes wrong when the matrix is altered after the first kernel execution. However, this problem also happens when the matrix is not an element of a matrix field...

Reply:
> in python scope, and they never appear in Taichi scope.

Yep, that's what `ensure_pyscope()` is for. I'm afraid I didn't really get your concern..

Reply:
I mean they never appear in Taichi scope (except passed as template arguments), so we don't need to check it.

Reply:
When they are passed as template arguments, they work well for the first time, and then the value in the kernel does not change when the values of the matrix changes, but this is the problem of template argument, not the problem of `Matrix`.

Reply:
> I mean they never appear in Taichi scope

If the code falls into this branch, sure. But note the entire function can still be called in Taichi scope. See for example: https://github.com/taichi-dev/taichi/blob/ecc0521fc506b6a5a4155d92f7642aa2c1701ffd/python/taichi/lang/matrix.py#L257-L259. (Note that this function is originally `__call__()`, which gets called in quite a few places in Taichi scope.) I'm adding this check to 1) make sure the function itself isn't miused, 2) as a better documentation to future developer who needs to touch this part.

Reply:
If we add `ensure_pyscope()` here, we ban using a `Matrix` in `MatrixField` as a template argument, but it shares the same problem with a normal python scope `Matrix` which is not banned.

Reply:
> > I mean they never appear in Taichi scope
> 
> If the code falls into this branch, sure. But note the entire function can still be called in Taichi scope. See for example:
> 
> https://github.com/taichi-dev/taichi/blob/ecc0521fc506b6a5a4155d92f7642aa2c1701ffd/python/taichi/lang/matrix.py#L257-L259
> . (Note that this function is originally `__call__()`, which gets called in quite a few places in Taichi scope.) I'm adding this check to 1) make sure the function itself isn't miused, 2) as a better documentation to future developer who needs to touch this part.

But when it is called inside Taichi scope, it does not cause trouble other than the old problem of template arguments. Why should we ban it? 
IIUC the only circumstance `self` can contain `SNodeHostAccess` is that a matrix in Matrix field is passed as a template argument?

Reply:
> Why should we ban it?

Because this just adds more confusion. `SNodeHostAccess` is just not supposed to exist in Taichi scope... If that happens, we should fix that 😂 

> it is a global variable?

Hmm.. how so?

Reply:
Like this.
```
import taichi as ti

ti.init()

a = ti.Matrix.field(n=2, m=2, dtype=ti.f32, shape=(10,))

b = a[0]

@ti.kernel
def foo():
    print(b[0, 0])

foo()
```

Reply:
As discussed, removed `ensure_pyscope`